### ISSUE TITLE
Nested schema deserialization depends on column order

### ISSUE THREAD

**Issue by ericphanson** [beacon-biosignals/Legolas.jl#94]
```julia
using Legolas
using Legolas: @schema, @version

@schema "test.issue" Issue

@version IssueV1 begin
    a::Int
    b::Int
end


@schema "test.parent" Parent
@version ParentV1 begin
    x::IssueV1
end


table = [ParentV1(; x=IssueV1(; a=1, b=2))]
Legolas.write("issue.arrow",table,  ParentV1SchemaVersion())

# New Julia session

using Legolas
using Legolas: @schema, @version

@schema "test.issue" Issue

@version IssueV1 begin
    b::Int
    a::Int
end


@schema "test.parent" Parent
@version ParentV1 begin
    x::IssueV1
end

Legolas.read("issue.arrow")

```
gives
```julia
julia> Legolas.read("issue.arrow").x[1]
IssueV1: (b = 1, a = 2)
```
which is incorrect, as `b` has value 1. This is because https://github.com/beacon-biosignals/Legolas.jl/blob/b86444ccc2b9f0cffa4f6fe8f8f724339fcc7ea4/src/schemas.jl#L649 assumes the order in the serialized file matches the existing one.

I am not sure if this is a bug or just a foot-gun (depending on what guarantees Legolas is supposed to have about column order not mattering).

**Comment by jrevels** [comment]
> I am not sure if this is a bug or just a foot-gun 

This is a (pretty bad 😨) bug

**Comment by jrevels** [comment]
The root of the problem here is that `Arrow.fromarrow` (and more specifically, the [Arrow.jl indexing code that invokes `fromarrow`](https://github.com/apache/arrow-julia/blob/787768fc60ff09ea52be0c5fb7c40e04be47619b/src/arraytypes/struct.jl#L46-L52)) splats accessed field values and does not pass the corresponding field keys through to the downstream constructor where they might be utilized; from https://docs.juliahub.com/ArrowTypes/3gg9u/2.2.2/autodocs/#ArrowTypes.fromarrow:

> StructKind: must overload fromarrow(::Type{T}, x...) where individual fields are passed as separate positional arguments; so if my custom type Interval has two fields first and last, then I'd overload like ArrowTypes.fromarrow(::Type{Interval}, first, last) = .... Note the default implementation is ArrowTypes.fromarrow(::Type{T}, x...) = T(x...), so if your type already accepts all arguments in a constructor no additional fromarrow method should be necessary (default struct constructors have this behavior).

So even though the Arrow-formatted table *does* encode the necessary field information to enable deserialization that is agnostic to the field ordering employed at serialization time, the interface that Arrow.jl provides to downstream type authors prevents that information from being utilized.

It'd be nice if the original `fromarrow` hadn't been designed this way, and that the relevant `getindex` method would instead simply pass through the whole `NamedTuple` encoded in the Arrow struct, instead of a splatted list of field values. I'm guessing the rationale here was performance (avoid accessing the field values that in most cases aren't needed).

Assuming that the reason really is performance, and does correspond to a real measurable perf difference in practice,
we might still be able to solve this in a non-breaking way by having ArrowTypes.jl define a trait like:

```
extract_field_names_upon_struct_getindex(T::Type) = false
```

then update the relevant Arrow.jl `getindex` method to check this trait; if `true`, it would extract and pass the whole `NamedTuple` through `fromarrow` instead of splatting the field values. then, type authors (and Legolas) could overload this trait to `true` for their type if they want Arrow.jl to pass through the whole `NamedTuple` to `fromarrow`.

Alternatively, if there's actually not a performance reason behind the original `fromarrow` splatted design, we could more simply define a passthrough method like:

```
fromarrow_struct(T::Type, x::NamedTuple) = fromarrow(T, values(x)...)
```

that would be a bit cleaner for `getindex` to invoke and for downstream callers to overload

**Comment by jrevels** [comment]
Using https://github.com/apache/arrow-julia/pull/493 in combination with this patch to Legolas:

```diff
diff --git a/src/schemas.jl b/src/schemas.jl
index 0ba6e26..675b16b 100644
--- a/src/schemas.jl
+++ b/src/schemas.jl
@@ -669,7 +669,7 @@ function _generate_record_type_definitions(schema_version::SchemaVersion, record
         $Arrow.ArrowTypes.ArrowType(::Type{R}) where {R<:$R} = NamedTuple{fieldnames(R),Tuple{fieldtypes(R)...}}
         $Arrow.ArrowTypes.toarrow(r::$R) = NamedTuple(r)
         $Arrow.ArrowTypes.JuliaType(::Val{$record_type_arrow_name}, ::Any) = $R
-        $Arrow.ArrowTypes.fromarrow(::Type{<:$R}, $(keys(record_fields)...)) = $R(; $(keys(record_fields)...))
+        $Arrow.ArrowTypes.fromarrow(::Type{<:$R}, nt::$Arrow.StructElement) = $R(; nt.fields...)
     end
```

resolves this issue. output of the test case from the OP with these fixes in place:

```
julia> Legolas.read("issue.arrow").x[1]
IssueV1: (b = 2, a = 1)
```

**Comment by jrevels** [comment]
will be resolved by https://github.com/beacon-biosignals/Legolas.jl/pull/106 once it is merged/released (and once https://github.com/apache/arrow-julia/pull/493 is released)

**Comment by jrevels** [comment]
resolved by #106
